Add ability to copy a dashboard

MarkRx opened the pull request:
Adds the ability to copy a dashboard - #7254. The dashboard and dashboard cards are duplicated but still refer to the same underlying questions.

#### Unanswered Questions:
* What record should be placed in the activity table when a copy is created? Should it be treated as as a dashboard-create and a dashboard-card-add? Or should there be a separate event for it? The current implementation only records a dashboard-create event (without any dashboard-card-add events).
* When copying do filter id's need to be regenerated? Currently filter fields are in json objects stored blobs between report_dashboard and report_dashboardcard and correlated by id. It seems sufficient that the id's are unique on a per dashboard basis but we are not sure.
* When a dashboard is copied what should the create and update dates be? We ended up going with the date the copy occurred.
* What icon should we used? I switched it from the copy icon to the clone icon. The pictures below do not reflect that change. 
![image](https://user-images.githubusercontent.com/17504314/47393243-ec4e9900-d6ec-11e8-8b06-e6b23feafb23.png)
* What verbiage should be used? Copy? Clone? Duplicate?
* How should the react history be stored for the copy dialog? If I copy a dashboard then hit the back button on a browser should I go back to the copy dialog or back to the original dashboard? The same question can be asked for the move dialog.


There are 3 ways to copy a dashboard
### 1. From the dashboard
![dashboard_copy_from_dashboard](https://user-images.githubusercontent.com/17504314/47045679-64531700-d161-11e8-8b24-d26c0124e263.PNG)

![image](https://user-images.githubusercontent.com/17504314/47390856-5879ce80-d6e6-11e8-8f02-bc1dbfdaf786.png)


### 2. From the collections landing page clicking on "..." (currently only available for dashboards)
This reuses the bulk copy dialog below
![collection_single_copy_dropdown](https://user-images.githubusercontent.com/17504314/47045554-14745000-d161-11e8-99c3-b82a3cc57beb.PNG)

### 3. As a bulk copy action (currently only available for dashboards)
![dashboard_multi_copy_dialog](https://user-images.githubusercontent.com/17504314/47045625-44bbee80-d161-11e8-8f22-798d769fad8e.PNG)
![dashboard_multi_copy](https://user-images.githubusercontent.com/17504314/47045651-530a0a80-d161-11e8-8265-b735c4519fbd.PNG)



###### Before submitting the PR, please make sure you do the following 
-  [x] If there are changes to the backend codebase, run the backend tests with `lein test && lein eastwood && lein bikeshed && lein docstring-checker && ./bin/reflection-linter`
-  [x] Run the frontend and integration tests with  `yarn lint && yarn flow && yarn test`) (can't get integration tests to run)
-  [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
(unless it's a tiny documentation change).

MarkRx committed:
Add UI for copying a dashboard

cgopalan committed:
Added tests for dashboard copy

MarkRx committed:
Run yarn prettier and fix lint errors

MarkRx committed:
Change copy icon to clone icon for copy dashboards

MarkRx committed:
Scale clone icon x2 so it aligns with sizes of other menu icons

MarkRx commented:
Switched to clone icon:

![image](https://user-images.githubusercontent.com/17504314/47438955-75131680-d779-11e8-96ad-3399202b65b1.png)

![image](https://user-images.githubusercontent.com/17504314/47438974-7e03e800-d779-11e8-8746-06f01ec3dc51.png)

I also think the edit icon on the dashboard should be the first icon after the +. Thoughts?

mazameli commented:
This is cool! A few thoughts after playing with this:

* > The dashboard and dashboard cards are duplicated but still refer to the same underlying questions.

To clarify, when duplicating, a new dashboard is created, but new cards are not created. The new dashboard points to the same saved questions.

* I would prefer `Duplicate` as the verb over `Copy`.
* The `clone` icon feels right to me, though @kdoh or I should probably clean up that asset.
* When copying a dashboard a dashboard from a collections list page, I think the modal that gets shown should be the "create" modal rather than the "move" modal, so that you can rename/describe the copy. Basically, show the same modal as currently gets shown when duplicating the dashboard while you're viewing the dashboard.
* To get this merged sooner, I think we should not include the duplicate action in the bulk actions bar at the bottom of the collections page when selecting multiple items. Reason being, I think it's a bit funky right now that because you can't yet duplicate questions or pulses, selecting a dashboard and a question/pulse disables the `Copy` button. We can add that bulk action back in if and when it's available for all entity types. (Bulk duplication is also a little hairy when it comes to the question of whether or not the user should be able to rename any/all duplicated items.)
* I can't intelligently answer your other open questions, so we'll need to rope in someone like @tlrobinson to take a look.

mazameli commented:
Another open question that just occurred to me is how this works with collection permissions. I.e., I should only be able to duplicate a dashboard into a location where I have Curate access, but what permission do I need to have on the originating dashboard? I _think_ I probably need Edit access on the collection the dashboard is in.

MarkRx committed:
Remove bulk copy; use copy entity dialog in collection landing; change UI 'copy' to 'duplicate'

MarkRx commented:
Changes:
- Removed bulk copy
- Used copy entity dialog
- Changed "copy" to "duplicate"
- Changed "Archive" to "Archive this item" for consistency

Open Questions:
1. If you copy a dashboard from the dashboard dialog it redirects you to a new page. If you then hit the back button on your browser it will attempt to take you back to the copy dialog. The problem is the dashboard is no longer in the redux state so the dialog won't have information about the dashboard (name and description for example). Since this didn't seem to fit I made the dashboard dialog rewrite history so as to remove the /copy route on a save. This however becomes inconsistent as hitting the back button after cancelling the dialog or using the move dialog will still open the modal.

What should be done? Should the history of using the move / copy modals be hidden? As a user I would expect that hitting the back button in the copy case will take me back to the browser (not back to the copy dialog). Likewise it seems strange to me that the move popup comes back after hitting the move button.

2. I think the edit icon should be the second icon in the list of icons in a dashboard (after the add card icon). It just feels wrong to me that the edit button is in the middle.
![image](https://user-images.githubusercontent.com/17504314/47578824-6878f480-d918-11e8-8e55-a4d8d19aa236.png)


![image](https://user-images.githubusercontent.com/17504314/47578499-93167d80-d917-11e8-8748-2a1d90cf6ddc.png)
![image](https://user-images.githubusercontent.com/17504314/47578526-9f023f80-d917-11e8-8721-b507471fbd2c.png)

mazameli commented:
This is looking quite great. To address your comments:

1. Yeah, I think you're actually right about the permissions.
2. I'm not sure what @tlrobinson thinks about addressing the Back behavior, but I agree it's weird with the Move modal. I think I'd propose not dealing with solving that in this PR, but we can peel that off as a separate issue since it's an existing problem with other modals.
3. I like your suggestion to make the `Edit` icon the second one.

Besides that, one nice bit of polish would be to add a confirmation dialog like we do when you move an item while in the collection view. The simplest version would be to have it say `Dashboard duplicated`. A step further would be that plus a link that says `See it`, which would take you to the url for the newly created dashboard.

Example of the move confirmation I'm talking about:

![screen shot 2018-10-29 at 3 34 35 pm](https://user-images.githubusercontent.com/2223916/47684628-c3b51c00-db90-11e8-805f-70b5caa3d634.png)

MarkRx committed:
Add copy toast for entities on collection landing page; reorder dashboard header icons

MarkRx committed:
Add .gitignore for vscode

MarkRx commented:
Changes:
- Added duplicate dashboard toast to collections landing. Message is "Duplicated dashboard" as that seems to be the verbiage order of other toasts. A "See It" link was also added. The icon corresponds to the entity model (dashboard, question, pulse). This differs from the collection move dialog which uses a checkbox (since it can copy multiple entities at once).
- Changed order of icons on dashboard header. The pencil is first followed by move and duplicate. I hid the move and duplicate icons when editing a dashboard since duplicating a dashboard while having unsaved changes would not be user friendly.

Notes:
- I didn't add a toast message when duplicating a dashboard within a dashboard. It didn't make sense to add since duplicating a dashboard redirects you to the new one.

![image](https://user-images.githubusercontent.com/17504314/47933949-62e95480-deac-11e8-8c6a-b897eb575727.png)

MarkRx committed:
Merge branch 'master' into copy-dashboard

mazameli commented:
This is looking really solid to me. @tlrobinson + @kdoh, could you guys find a moment to give this a review?

kdoh reviewed (approved):
This looks and works great. Thanks for the work @MarkRx!

senior commented on src/metabase/api/dashboard.clj:
```diff
@@ -188,6 +188,46 @@
+                           (u/prog1 (db/insert! Dashboard dashboard-data)
+                                    ;; Get cards from existing dashboard and associate to copied dashboard
+                                    (doseq [card (existing-dashboard :ordered_cards)]
+                                      (u/prog1 (api/check-500 (dashboard/add-dashcard! <> (card :card_id) card))))))]
```
I think this extra `u/prog1` isn't needed

senior commented on test/metabase/api/dashboard_test.clj:
```diff
@@ -543,6 +543,28 @@
+          (-> ((user->client :rasta) :post 200 (format "dashboard/%d/copy" dashboard-id))
+              (dissoc :id)
+              (update :created_at boolean)
+              (update :updated_at boolean))))
```
You can use `metabase.test.util/boolean-ids-and-timestamps` instead of the above three lines, same with line 563-565 as well.

senior commented on src/metabase/api/dashboard.clj:
```diff
@@ -188,6 +188,46 @@
   (update dashboard :ordered_cards add-query-average-duration-to-dashcards))
 
 
+(defn get-dashboard
```
This isn't used outside of the namespace so it can be marked as private via `defn-`

senior reviewed (changes requested):
Thanks for you contribution @MarkRx. When running the tests I found two test failures:
```
failure in (dashboard_test.clj:546) : metabase.api.dashboard-test
failure in (dashboard_test.clj:556) : metabase.api.dashboard-test
```
Were you seeing those when you ran the tests locally?

MarkRx committed:
Clojure PR changes; Unit test improvements

senior commented on src/metabase/api/dashboard.clj:
```diff
@@ -188,6 +188,46 @@
+  "Get `Dashboard` with ID."
+  [id]
+  (-> (Dashboard id)
+               api/check-404
```
Indentation is off here

senior commented on src/metabase/api/dashboard.clj:
```diff
@@ -188,6 +188,46 @@
+  ;; if we're trying to save the new dashboard in a Collection make sure we have permissions to do that
+  (collection/check-write-perms-for-collection collection_id)
+  (let [
+      existing-dashboard (get-dashboard from-dashboard-id)
```
This should be moved up a line

senior commented on src/metabase/api/dashboard.clj:
```diff
@@ -188,6 +188,46 @@
+  (collection/check-write-perms-for-collection collection_id)
+  (let [
+      existing-dashboard (get-dashboard from-dashboard-id)
+      dashboard-data {:name                  (or name (:name existing-dashboard))
```
Indentation and spacing are off here.

senior commented on src/metabase/api/dashboard.clj:
```diff
@@ -188,6 +188,46 @@
+                          ;; Ok, now save the Dashboard
+                          (u/prog1 (db/insert! Dashboard dashboard-data)
+                                  ;; Get cards from existing dashboard and associate to copied dashboard
+                                  (doseq [card (existing-dashboard :ordered_cards)]
```
Missed flipping this to using the keyword as the function. Same with `:card_id` below.

senior reviewed (dismissed):
Looks good, just some whitespace/indentation issues to fixup.

MarkRx committed:
PR tweaks

MarkRx commented:
Should be good now. @cgopalan

camsaul reviewed (approved):
Looks good to me

kdoh commented:
I think it'd be a nice addition.

mazameli commented:
@tlrobinson Bulk copy was explicitly removed because of the weirdness in it becoming enabled/disabled if items other than dashboards are selected ([see this comment](https://github.com/metabase/metabase/pull/8754#issuecomment-432816574)), with the plan being to bring that back once copying is enabled for all selectable entities.

tlrobinson commented:
Nevermind, I see bulk copy was removed.

tlrobinson commented:
Looks like there's some bikeshed linter errors:

```
Badly formatted files:
paychex-copy-dashboard/test/metabase/api/dashboard_test.clj:571:            (dashboard-response ((user->client :crowberto) :post 200 (format "dashboard/%d/copy" dashboard-id)
paychex-copy-dashboard/test/metabase/api/dashboard_test.clj:576:(expect
paychex-copy-dashboard/test/metabase/api/dashboard_test.clj:584:      (count (db/select-ids DashboardCard, :dashboard_id
paychex-copy-dashboard/test/metabase/api/dashboard_test.clj:585:        (u/get-id
paychex-copy-dashboard/test/metabase/api/dashboard_test.clj:597:        (= (db/select-one-field :collection_id Dashboard :id
```

I don't really know what format bikeshed expects. @camsaul?

MarkRx committed:
Fix some lint whitespace issues

camsaul commented:
oh ok, makes sense. Thanks

I'll do a final review in a second and merge if everything checks out

camsaul commented:
Merged in by #8975

Thanks!
